### DISCUSSION TITLE
Best way to extend cds-button

### DISCUSSION THREAD

**Discussion by wokis** [vmware-clarity/core#14] [Q&A]
The old clr angular buttons had a briefly check mark shown when set to `ClrLoadingState.SUCCESS`. The new `cds-button` are stateless and I was wondering which way is the best way to extend `cds-button` so that it will behave like the old angular buttons do.

So far I've come up with:
```typescript
<cds-button (click)="submit()" [loadingState]="submitButtonState"></cds-button>

submit() {
  this.submitButtonState = 'success';
  setTimeout(() => {
    this.submitButtonState = 'default';
  }, 2000);
}
```

But I would have to add it to every `submit()` method in my application. Is it possible to extend `cds-button` with this behavior? I've been looking at Angular directives but I must say I'm not sure it's the best way or if at all possible to extend Clarity Core components this way.

**Comment by mathisscott** [comment]
There are a couple of paths forward for you here. One would be creating a component at the form level in Angular that could be reused throughout your application. This could be built in a way such that it can handle/abstract this behavior. In theory, you could also create a custom directive that you could then add to existing forms without the need for a separate component abstraction.

Another option would be to create a custom Angular component with the button. This could wrap (or extend) the existing component and you could trigger the behavior you want at the button level. This would still require a change to your forms because that behavior would have to be initiated from somewhere.

All in all, I think a custom directive that you can add to your form tags is probably the most straightforward option.

**Reply by wokis** [reply]
So far I've come up with:
```typescript
@Directive({
  selector: 'cds-button'
})
export class TildaCdsButtonDirective extends CdsButtonDirective {
  set loadingState(value: any) {
    super.loadingState = value;
    if (value == ClrLoadingState.success || value == ClrLoadingState.error) {
      setTimeout(() => {
        super.loadingState = ClrLoadingState.default;
      }, 600);
    }
  }
}
```
That takes care of the loading state but the transition animation from the clr buttons are still lost. I haven't found a good way to animate the `cds-icon` inside the button from a directive.

What was the reasoning for dropping the button state transition animations in cds core?

**Reply by mathisscott** [reply]
I don't think that dropping the button state transition was intentional. At the time the button in `@cds/core` was built, we had not created the animation library. We now have the animation library in `@cds/core` so it would be pretty straightforward to introduce the success animation again. 

I will open an issue to do that.

**Reply by mathisscott** [reply]
Github issue for the button animation: https://github.com/vmware-clarity/core/issues/16
